src/main/java/com/configcat/ConfigService.java

```diff
@@ -0,0 +1,264 @@
+        } else {
+            return fetchIfOlder(Constants.DISTANT_PAST, true)
+                    .thenApply(entryResult -> {
+                        return new SettingResult(entryResult.value().getConfig().getEntries(), entryResult.value().getFetchTime());
+                    });
+        }
+
+    }
+
+    private CompletableFuture<Result<Entry>> fetchIfOlder(long time, boolean preferCached) {
+        lock.lock();
+        try {
+            // Sync up with the cache and use it when it's not expired.
+            if (cachedEntry.isEmpty() || cachedEntry.getFetchTime() > time) {
+                Entry fromCache = readCache();
+                if (!fromCache.isEmpty() && !fromCache.getETag().equals(cachedEntry.getETag())) {
+                    cachedEntry = fromCache;
+                }
+                // Cache isn't expired
+                if (cachedEntry.getFetchTime() > time) {
+                    initialized = true;
+                    return CompletableFuture.completedFuture(Result.success(cachedEntry));
+                }
+            }
+            // Use cache anyway (get calls on auto & manual poll must not initiate fetch).
+            // The initialized check ensures that we subscribe for the ongoing fetch during the
+            // max init wait time window in case of auto poll.
+            if (preferCached && initialized) {
+                return CompletableFuture.completedFuture(Result.success(cachedEntry));
+            }
+            // If we are in offline mode we are not allowed to initiate fetch.
+            if (offline.get()) {
+                return CompletableFuture.completedFuture(Result.success(cachedEntry));
+            }
+
+            if (runningTask == null) {
+                // No fetch is running, initiate a new one.
+                runningTask = new CompletableFuture<>();
+                configFetcher.fetchAsync(cachedEntry.getETag())
+                        .thenAccept(this::processResponse);
+            }
+
+            return runningTask;
+
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private synchronized void broadcastConfigurationChanged() {
+        if (this.listeners != null) {
+            for (ConfigurationChangeListener listener : this.listeners) {
+                listener.onConfigurationChanged();
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (!this.closed.compareAndSet(false, true)) {
+            return;
+        }
+        if (pollingMode instanceof AutoPollingMode) {
+            if (pollScheduler != null) this.pollScheduler.shutdown();
+            if (initScheduler != null) this.initScheduler.shutdown();
+            if (listeners != null) this.listeners.clear();
+        }
+        this.configFetcher.close();
+    }
+
+    public void setOnline() {
+        if (closed.get()) {
+            logger.warn("Client has already been closed, this 'setOnline' has no effect.");
```

Comment:
```suggestion
            logger.warn("The 'setOnline' method has no effect because the client has already been closed");
```

Reply:
@z4kn4fein , @sigewuzhere  To be honest, I would remove these checks from here.  The close checks already happening in the ConfigCatClient setOnline and setOffline methods, also the message says 'has no effect' but there is no return here. So we log the message and we run the actual code ... not nice.

Reply:
I agree. It doesn't necessary to do the same check here again.

Reply:
Agreed